### PR TITLE
Fetch witnesses and divulgences in the same query

### PR DESCRIPTION
**Pull request by gerolf-da**
We use postgres' array_agg function to fetch witnesses and divulgences
for a contract in the same query, instead of fetching the contract first
and then doing another 2 db roundtrips to fetch witnesses and
divulgences.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala [hunk 1]
```diff
@@ -718,23 +716,17 @@ private class PostgresLedgerDao(
...
-    SQL("select witness from contract_witnesses where contract_id={contract_id}")
-
-  private val DivulgenceParser = (party("party")
-    ~ long("ledger_offset")
-    ~ ledgerString("transaction_id") map (flatten))
-
-  private val SQL_SELECT_DIVULGENCE =
-    SQL(
-      "select party, ledger_offset, transaction_id from contract_divulgences where contract_id={contract_id}")
+      "select c.*, le.effective_at, array_agg(witness) witnesses, coalesce(array_agg(array[d.party, d.transaction_id]) filter (where d.party is not null), array[]::varchar[][]) divulgences from contracts c inner join ledger_entries le on c.transaction_id = le.transaction_id inner join contract_witnesses w on c.id = w.contract_id left join contract_divulgences d on c.id = d.contract_id where id={contract_id} and archive_offset is null group by c.id, le.effective_at")
```
Comment (gaboraranyossy-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: pretty cool!

**Comment by gerolf-da** [comment]
@rautenrieth-da: we noticed that a gazillion queries are executed when a client fetches the active contracts, that is: up to 3 additional queries with separate network roundtrips.

And now that I'm playing around with `array_agg`, I'm wondering why we don't just store the key maintainers, witnesses, and divulgence either as arrays directly in the table, or maybe as a custom data type?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala [hunk 2]
```diff
@@ -718,26 +718,36 @@ private class PostgresLedgerDao(
...
+        |   le.effective_at,
+        |   array_agg(witness) witnesses,
+        |   coalesce(array_agg(array[d.party, d.transaction_id]) filter (where d.party is not null), array[]::varchar[][]) divulgences,
+        |   case when c.key is not null
+        |        then array(select m.maintainer from contract_key_maintainers m where m.contract_id = c.id)
+        |	       else array[]::varchar[]
+        |   end maintainers
+        |from contracts c
+        |   inner join ledger_entries le
+        |       on c.transaction_id = le.transaction_id
```
Comment (dajmaki) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: Just a heads up: We should depend on as little as possible on the ledger entries table and store all required data in the "ACS" tables to be prepared to support participant index pruning (time-based trimming of the ledger entries table).

Comment (dajmaki) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala]: Another thing we should soon fix is being able to insert divulged contracts into the index, which obviously won't have associated transaction (we wouldn't even know the transaction id). Would probably be good to get together and discuss how the database schema needs to evolve to work as the index database for a participant with partial view on the ledger...

**Pull request closed without merging**